marimbabot_speech/src/speech2text.py

```diff
@@ -1,84 +1,109 @@
 
+# Speech to text recognition, it is a wrapper of whisper
 class STT:
 	def __init__(self):
+		self.prompt = self.generate_prompt()
+		self.init_audio_config()
+		self.init_whisper_model()
+		self.warm_up()  # warm up the whisper model
+		self.init_ros_node()
+
+	def init_audio_config(self):
+		self.sr = rospy.get_param('~sr', 16000)
+
+	def init_whisper_model(self):
 		# 'tiny.en', 'tiny', 'base.en', 'base', 'small.en', 'small', 'medium.en', 'medium', 'large-v1', 'large-v2', 'large'
 		self.model = whisper.load_model('medium.en')
-		self.file_controller = WAVFile()
-		self.speech_pub = rospy.Publisher('/speech_node/speech', SpeechMsg, queue_size=100, tcp_nodelay=True)
-		self.tmp_sub = rospy.Subscriber('/speech_node/audio_tmp', TmpFileMsg, self.tmp_callback, queue_size=10, tcp_nodelay=True)
-		self.recognize_freq = 2  # Hz
-		self.recognize_rate = rospy.Rate(self.recognize_freq)
-		self.no_speech_prob_filter = 0.5
-		rospy.logdebug(f"cwd:{os.getcwd()}")
-		self.prompt = self.generate_prompt()
 
-	def generate_prompt(self):
-		commands = get_commands()
-		base_prompt = '''Marimbabot is a marimba playing robot arm. The possible commands include:		
-		'''
-		for command in commands:
-			base_prompt += ''.join(f'{command.strip()}, ')
-		rospy.logdebug(f"Generate prompt as: {base_prompt}.")
-		return base_prompt
+	def warm_up(self):
+		# 30 denote sec, since whisper take fixed length of audio data as input
+		self.recognize(np_data=np.zeros(self.sr*30, dtype=np.float32),pub_speech=False)
+		rospy.loginfo("Whisper model is warmed up!")
 
+	def init_ros_node(self):
+		self.buffer_sub = rospy.Subscriber(
+			'/speech_node/speech_buffer',
+			AudioData,
+			self.handle_audio_buffer,
+			queue_size=100,
+			tcp_nodelay=True)
 
-	def tmp_callback(self, tmp_file_msg):
-		file_path = tmp_file_msg.file_path
-		text, no_speech_prob = self.recognize(file_path)
-		if no_speech_prob > self.no_speech_prob_filter:
-			return
-		if len(text)>300:
-			return
-		speech_msg = SpeechMsg()
-		speech_msg.header.stamp = rospy.Time.now()
-		speech_msg.speech = text
-		speech_msg.is_finished = tmp_file_msg.is_finished
-		self.speech_pub.publish(speech_msg)
-		rospy.logdebug(f"speech published.")
+		# To publish the recognized text
+		self.speech_pub = rospy.Publisher(
+			'speech',
+			SpeechMsg,
+			queue_size=100,
+			tcp_nodelay=True)
 
+	def handle_audio_buffer(self, req):
+		text, no_speech_prob = self.recognize(np_data=self.unpack_stream(req.data))
+		msg = SpeechMsg()
+		msg.header.stamp = rospy.Time.now()
+		msg.text = text
+		msg.no_speech_prob = no_speech_prob
+		self.speech_pub.publish(msg)
 
-	def recognize(self, file_path:str):
-		time_0 = time.time()
-		# load audio and pad/trim it to fit 30 seconds
-		audio = whisper.load_audio(file_path)
-		audio = whisper.pad_or_trim(audio)
+	# To address the way how to unpack the data for whisper
+	def unpack_stream(self, data):
+		return np.array(struct.unpack(f"{int(len(data) / 2)}h", bytes(data)), dtype=float) / 526
```

Comment:
This can be done more efficiently by constructing a numpy dtype that matches the struct and then just call np.frombuffer irc.

Reply:
![image](https://github.com/UHHRobotics22-23/marimbabot/assets/15075613/85e12908-f35d-4f5f-951b-2999c80f1a22)

See the speedup and it is also simpler.

Reply:
You probably also don't need the `bytes(data)` cast

Reply:
As I remember, I have tried your solution, and the buffer has not successfully been interpreted. But not pretty sure of it, I will try again and then tell you the result.

Reply:
done